pyro/infer/mcmc/trace_kernel.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import absolute_import, division, print_function
+
+from abc import abstractmethod, ABCMeta
+
+from six import add_metaclass
+
+
+@add_metaclass(ABCMeta)
+class TraceKernel(object):
+
+    def setup(self, *args, **kwargs):
+        """
+        Optional method to set up any state required before the run.
+
+        :param \*args: Algorithm specific positional arguments.
+        :param \*\*kwargs: Algorithm specific keyword arguments.
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Optional method to clean up any residual state on termination.
+        """
+        pass
+
+    def diagnostics(self, time_step):
+        """
+        Relevant diagnostics (optional) to be printed at regular intervals
+        of the MCMC run. Returns `None` by default.
+
+        :param time_step: Current Monte Carlo time-step.
+        :return: String containing the diagnostic summary. e.g. acceptance rate
+        :rtype: string
+        """
+        return None
+
+    @abstractmethod
+    def sample(self, trace, time_step, *args, **kwargs):
```

Comment:
Do you have a reason for hard-coding `time_step` rather than subsuming it in `*args, **kwargs`?

Reply:
The only class that currently calls this is the wrapper `MCMC` class which passes this while calling the kernel's sample method (though it is arguable whether we need to do this or have the kernel class track this as part of its state as @eb8680 mentioned). I am thinking of removing this and have the kernel instance track it if it needs to. Then we can remove it from the method signature as well.